src/xrpld/overlay/detail/OverlayImpl.cpp

```diff
@@ -1199,19 +1200,35 @@ OverlayImpl::getManifestsMessage()
 void
 OverlayImpl::relay(
     uint256 const& hash,
-    protocol::TMTransaction& m,
+    std::optional<std::reference_wrapper<protocol::TMTransaction>> tx,
     std::set<Peer::id_t> const& toSkip)
 {
-    auto const sm = std::make_shared<Message>(m, protocol::mtTRANSACTION);
+    std::shared_ptr<Message> sm = nullptr;
+    bool relay = tx.has_value();
+    if (relay)
+    {
+        auto& txn = tx->get();
+        SerialIter sit(makeSlice(txn.rawtransaction()));
+        relay = !isPseudoTx(STTx{sit});
+    }
+    if (relay)
+    {
+        auto& txn = tx->get();
+        sm = std::make_shared<Message>(txn, protocol::mtTRANSACTION);
+    }
+
+    if (!relay && !app_.config().TX_REDUCE_RELAY_ENABLE)
+        return;
```

Comment:
Just kind of putting the changes into context from @bthomee's comment at https://github.com/XRPLF/rippled/pull/5243/files#r1929171355. I think this is safe, no risk of runtime error, then you can get rid of all the `relay` checks after this point.
```suggestion
    std::shared_ptr<Message> sm = nullptr;
    bool relay = tx.has_value();
    if (relay)
    {
        auto& txn = tx->get();
        SerialIter sit(makeSlice(txn.rawtransaction()));
        relay = !isPseudoTx(STTx{sit});
    }

    if (!relay) {
        if (!app_.config().TX_REDUCE_RELAY_ENABLE)
            return;
        auto peers = getActivePeers(toSkip, 0, 0, 0)
        JLOG(journal_.trace()) << "not relaying tx, total peers " << peers.size();
        for (auto const& p : peers)
            p->addTxQueue(hash);
        return;
   }

    auto& txn = tx->get();
    sm = std::make_shared<Message>(txn, protocol::mtTRANSACTION);
```

Reply:
if we do that, then we should refactor further down:

```
for (auto const& p : peers)
    {
        // always relay to a peer with the disabled feature
        if (!p->txReduceRelayEnabled())
        {
            if (relay)
                p->send(sm);
        }
        else if (enabledAndRelayed < enabledTarget)
        {
            assert(relay);
            enabledAndRelayed++;
            p->send(sm);
        }
        else
        {
            p->addTxQueue(hash);
        }
    }
```

1) check for relay is not needed //  if (relay)
2) last else is also might not be need (not sure about that) // p->addTxQueue(hash);

Reply:
> if we do that, then we should refactor further down:

Like I said, you can get rid of all the relay checks further down.

The last else (`p->addTxQueue(hash)`) is still needed because that's part of the reduce relay logic that will still be invoked if the transaction is relay-able.